lib/SyTest/Homeserver/ProcessManager.pm

```diff
@@ -237,5 +238,129 @@ sub _kill_process
    );
 }
 
-1;
+=head2 _start_process_and_await_notify
+
+   $fut = $hs->_start_process_and_await_notify( %params )
+
+This method starts a new process, setting the `NOTIFY_SOCKET` env, and waits to
+receive a `READY=1` notification from the process on the socket.
+
+Parameters are passed on to C<IO::Async::Process::new>.
+
+=cut
+
+sub _start_process_and_await_notify
+{
+   my $self = shift;
+
+   my %params = @_;
+
+   # We now need to do some faffing to pull out any passed env hash so that we
+   # can then pass it to `_await_ready_notification`.
+   #
+   # The env is passed in as part of the `setup` param, which is an ordered map
+   # represented as an array.
+   $params{setup} //= [];
+
+   my $setup = $params{setup};
+
+   my %setup_map = @$setup;  # Copy to a hash so we can pull out the env entry.
+   my $env = $setup_map{env} // {};
+   if ( not defined $setup_map{env} ) {
+      # There was no env entry, so we me need to add it to the setup array.
+      push @$setup, env => $env;
+   }
+
+   # We need to set this up *before* we start the process as we need to bind the
+   # socket before startin the process.
+   my $await_fut = $self->_await_ready_notification( $env );
+
+   my $proc = $self -> _start_process( %params );
+   my $finished_future = $self->{proc_info}{$proc}->finished_future;
+
+   my $fut = Future->wait_any(
+      $await_fut,
+      $finished_future->without_cancel()->then_fail(
+         "Process died without becoming connectable",
+      ),
+   )->else_with_f( sub {
+      my ( $f ) = @_;
+
+      # We need to manually kill child procs here as we don't seem to have
+      # registered the on finish handler yet.
+      $self->kill_and_await_finish()->then( sub {
+         $f
+      })
+   } );
+   return $fut;
+}
+
+=head2 _await_ready_notification
+
+   $fut = $hs->_await_ready_notification( $env )
+
+This method binds a listener to a newly created unix socket and waits for a
+`READY=1` to be received. The socket address is added to the `env` map passed
+in under `NOTIFY_SOCKET`.
+
+This is basically a noddy implementation of the `sd_notify` mechanism.
+
+=cut
+
+sub _await_ready_notification
+{
+   my $self = shift;
+
+   my ( $env ) = @_;
 
+   my $loop = $self->loop;
+   my $output = $self->{output};
+
+   # Create a random abstract socket name. Abstract sockets start with a null
+   # byte.
+   my $random_id = join "", map { chr 65 + rand 25 } 1 .. 20;
+   my $path = "\0sytest-$random_id.sock";
+
+   # We replace null byte with '@' to allow us to pass it in via env. (This is
+   # as per the sd_notify spec).
+   $env->{"NOTIFY_SOCKET"} = $path =~ s/\0/\@/rg;
```

Comment:
cleaner to do this outside `_await_ready_notification` (possibly in another helper function) and just pass in the `$path` to `_await_ready_notification` ?

Reply:
I had a look at this, and basically either a) you move it to a function that accepts `$env` which I don't think buys you very much, or b) move this block of code up to the `_start_process_and_await..`  function which just makes it a bit messy.

I kind of like have all the sd notify bits in one place tbh. Maybe it'd be better to rename it `_create_socket_and_await_notification` or something?

Reply:
well, maybe. whatever.